ml/functions.py

```diff
@@ -0,0 +1,43 @@
+from typing import List
+
+import torch
+from torch import Tensor
+
+from ml.typing import TensorWithMask
+
+
+def concat(tensors: List[TensorWithMask], dim):
+    embeddings = []
+    masks = []
+
+    for tensor_with_mask in tensors:
+        embeddings.append(tensor_with_mask.tensor)
+        masks.append(tensor_with_mask.mask)
+
+    return TensorWithMask(
+        torch.cat(embeddings, dim=dim),
+        torch.cat(masks, dim=dim)
+    )
+
+
+def softmax(inputs: TensorWithMask, dim) -> Tensor:
+    scores = inputs.tensor
+    mask = inputs.mask
+
+    scores[~mask] = -1e18  # TODO [Vladimir Baikalov]: use torch INF here
+    probabilities = torch.softmax(scores, dim=dim)
+    return probabilities
+
+
+def euclidean_distance(left_tensor: TensorWithMask, right_tensor: Tensor, dim) -> TensorWithMask:
```

Comment:
It's better to have `right_tensor` and `dim` variables with types `TensorWithMask`  and `int` types respectively.
```suggestion
def euclidean_distance(left_tensor: TensorWithMask, right_tensor: TensorWithMask, dim: int) -> TensorWithMask:
```